File: scripts/fix_example_sentences.py
#!/usr/bin/env python3
"""
Transform 267 vocabulary files:
  - Extract inline example blocks (JP / <small>*romaji*</small> / EN)
  - Reformat as ## Example Sentences with keigo-card blue div
  - Place the new section just before ## Related Expressions
"""

import re
import os
import glob

EXAMPLE_RE = re.compile(
    r'^([^\n<\-#*][^\n]*)\n'          # JP line (not starting with special chars)
    r'<small>\*([^*\n]+)\*</small>\n'  # <small>*romaji*</small>
    r'([^\n]+)',                        # EN line
    re.MULTILINE,
)

def build_keigo_card(examples):
    items = []
    for jp, romaji, en in examples:
        items.append(
            f'<span class="keigo-jp"><code>{jp.strip()}</code></span><br>\n'
            f'<em>{romaji.strip()}</em><br>\n'
            f'{en.strip()}'
        )
    return (
        '## Example Sentences\n\n'
        '<div class="keigo-card blue">\n'
        + '\n<br><br>\n'.join(items)
        + '\n</div>'
    )

def transform_file(filepath):
    with open(filepath, 'r', encoding='utf-8-sig') as f:
        raw = f.read()

    if '## Example Sentences' in raw:
        return 'skip'
    if '<small>' not in raw:
        return 'no_inline'
    if '## Related Expressions' not in raw:
        return 'no_related'

    split_idx = raw.index('## Related Expressions')
    body = raw[:split_idx]
    tail = raw[split_idx:]

    examples = EXAMPLE_RE.findall(body)
    if not examples:
        return 'no_match'

    # Remove inline blocks from body
    cleaned = EXAMPLE_RE.sub('', body)
    cleaned = re.sub(r'\n{3,}', '\n\n', cleaned).rstrip() + '\n\n'

    keigo = build_keigo_card(examples)
    new_content = cleaned + keigo + '\n\n' + tail

    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(new_content)

    return f'ok ({len(examples)} examples)'

def main():
    files = sorted(glob.glob('content/vocabulary/**/*.md', recursive=True))

    counts = {}
    for filepath in files:
        with open(filepath, 'r', encoding='utf-8-sig') as f:
            if '## Example Sentences' in f.read():
                continue
        result = transform_file(filepath)
        counts[result] = counts.get(result, 0) + 1
        if not result.startswith('ok'):
            print(f'  [{result}] {filepath}')

    print('\n--- Summary ---')
    for k, v in sorted(counts.items()):
        print(f'  {k}: {v}')

if __name__ == '__main__':
    main()

File: scripts/fix_example_sentences.js
#!/usr/bin/env node
/**
 * Transform vocabulary files missing ## Example Sentences:
 *   - Extract inline blocks (JP / <small>*romaji*</small> / EN)
 *   - Reformat as keigo-card blue div under ## Example Sentences
 *   - Insert just before ## Related Expressions
 */

const fs   = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Match: JP line \n <small>*romaji*</small> \n EN line
const EXAMPLE_RE = /^([^\n<\-#*][^\n]*)\n<small>\*([^*\n]+)\*<\/small>\n([^\n]+)/gm;

function buildKeigoCard(examples) {
  const items = examples.map(([jp, romaji, en]) =>
    `<span class="keigo-jp"><code>${jp.trim()}</code></span><br>\n` +
    `<em>${romaji.trim()}</em><br>\n` +
    en.trim()
  );
  return (
    '## Example Sentences\n\n' +
    '<div class="keigo-card blue">\n' +
    items.join('\n<br><br>\n') +
    '\n</div>'
  );
}

function transformFile(filepath) {
  const raw = fs.readFileSync(filepath, 'utf8').replace(/^\uFEFF/, ''); // strip BOM

  if (raw.includes('## Example Sentences')) return 'skip';
  if (!raw.includes('<small>'))              return 'no_inline';
  if (!raw.includes('## Related Expressions')) return 'no_related';

  const splitIdx = raw.indexOf('## Related Expressions');
  const body = raw.slice(0, splitIdx);
  const tail = raw.slice(splitIdx);

  const examples = [];
  let m;
  EXAMPLE_RE.lastIndex = 0;
  while ((m = EXAMPLE_RE.exec(body)) !== null) {
    examples.push([m[1], m[2], m[3]]);
  }

  if (examples.length === 0) return 'no_match';

  // Remove inline blocks from body
  EXAMPLE_RE.lastIndex = 0;
  let cleaned = body.replace(EXAMPLE_RE, '');
  cleaned = cleaned.replace(/\n{3,}/g, '\n\n').trimEnd() + '\n\n';

  const keigo = buildKeigoCard(examples);
  const newContent = cleaned + keigo + '\n\n' + tail;

  fs.writeFileSync(filepath, newContent, 'utf8');
  return `ok (${examples.length} examples)`;
}

// Collect files missing ## Example Sentences
function getVocabFiles(dir) {
  const results = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) results.push(...getVocabFiles(full));
    else if (entry.name.endsWith('.md')) results.push(full);
  }
  return results.sort();
}

const vocabDir = path.join(__dirname, '..', 'content', 'vocabulary');
const allFiles = getVocabFiles(vocabDir);
const targets  = allFiles.filter(f => !fs.readFileSync(f, 'utf8').includes('## Example Sentences'));

console.log(`Files to transform: ${targets.length}`);

const counts = {};
for (const fp of targets) {
  const result = transformFile(fp);
  counts[result] = (counts[result] || 0) + 1;
  if (!result.startsWith('ok')) {
    console.log(`  [${result}] ${fp}`);
  }
}

console.log('\n--- Summary ---');
for (const [k, v] of Object.entries(counts).sort()) {
  console.log(`  ${k}: ${v}`);
}
